### PR TITLE
finalize `--rocksdb.sync-interval` feature

### PR DESCRIPTION
**Pull request by jsteemann**
* added configuration option `--rocksdb.sync-interval`

  This option specifies interval (in milliseconds) that ArangoDB will use to 
  automatically synchronize data in RocksDB's write-ahead log (WAL) files to 
  disk. Automatic syncs will only be performed for not-yet synchronized data, 
  and only for operations that have been executed without the *waitForSync* 
  attribute.

  Automatic synchronization is performed by a background thread. The default
  sync interval is 100 milliseconds.

### CODE REVIEW

File: arangod/RocksDBEngine/RocksDBEngine.cpp [hunk 1]
```diff
@@ -619,6 +662,16 @@ void RocksDBEngine::stop() {
     }
     _backgroundThread.reset();
   }
+  
+  if (_syncThread) {
+    _syncThread->beginShutdown();
+
+    // wait until sync thread stops
+    while (_syncThread->isRunning()) {
+      std::this_thread::sleep_for(std::chrono::microseconds(10000));
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBEngine.cpp]: wouldn't `std::this_thread::yield()` be better ?

File: arangod/RocksDBEngine/RocksDBSyncThread.cpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+#include "Basics/RocksDBUtils.h"
+#include "Logger/Logger.h"
+#include "RocksDBEngine/RocksDBEngine.h"
+
+#include <rocksdb/status.h>
+#include <rocksdb/utilities/transaction_db.h>
+
+using namespace arangodb;
+
+RocksDBSyncThread::RocksDBSyncThread(RocksDBEngine* engine, uint64_t interval)
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBSyncThread.cpp]: you should use `std::chrono::milliseconds` for the parameter

File: arangod/RocksDBEngine/RocksDBSyncThread.cpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+     
+      {
+        // wait for time to elapse, and after that update last sync time 
+        CONDITION_LOCKER(guard, _condition);
+
+        auto const previousLastSequenceNumber = _lastSequenceNumber;
+        auto const previousLastSyncTime = _lastSyncTime;
+        auto const end = _lastSyncTime + std::chrono::microseconds(_interval);
+        if (end > now) {
+          guard.wait(std::chrono::microseconds(std::chrono::duration_cast<std::chrono::microseconds>(end - now)));
```
Comment (graetzer) [inline arangod/RocksDBEngine/RocksDBSyncThread.cpp]: should you check `isStopping()` after waiting here ?

**Comment by jsteemann** [comment]
http://jenkins01.arangodb.biz:8080/view/PR/job/arangodb-matrix-pr/297/

File: arangod/RocksDBEngine/RocksDBSyncThread.cpp (hunk 3 above)
Comment (jsteemann) [reply arangod/RocksDBEngine/RocksDBSyncThread.cpp]: I don't think this is required.
If we `continue`, then we will directly afterwards check `isStopping()` anyway.
If we don't `continue`, we will do one final sync and then check `isStopping()` afterwards.
I don't think the extra sync on shutdown will do any harm (it's probably even sensible to do it...).

**Pull request merged**